lib/bigtable/row.js

```diff
@@ -0,0 +1,620 @@
+
+  chunks
+    .reduce(function(chunkList, chunk) {
+      if (chunk.resetRow) {
+        return [];
+      }
+
+      if (chunk.rowContents) {
+        chunkList.push(chunk.rowContents);
+      }
+
+      return chunkList;
+    }, [])
+    .forEach(function(chunk) {
+      var family = chunk.name;
+
+      rowData[family] = rowData[family] || {};
+
+      chunk.columns.forEach(function(column) {
+        var qualifier = Entry.convertFromBytes(column.qualifier);
+
+        column.cells.forEach(function(cell) {
+          var value = Entry.convertFromBytes(cell.value);
+          rowData[family][qualifier] = value;
+        });
+      });
+    });
+
+  return rowData;
+};
+
+/**
+ * Batching allows you to perform multiple operations in a single request.
+ * When you are finished making your changes, run the `done()` function provided
+ * in the callback function to complete the mutations. See an example below for
+ * more information.
+ *
+ * @param {function} fn - The function to create a batch request.
+ * @param {function} fn.done - Function to be called to finalize request.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {object} callback.apiResponse - The full API response.
+ *
+ * @example
+ * row.batch(function(done) {
+ *   row
+ *     .save('name:first', 'Abe')
+ *     .deleteColumns('life');
+ *
+ *   done();
+ * }, function(err, apiResponse));
+ */
+Row.prototype.batch = function(fn, callback) {
+  if (!callback) {
+    var row = extend(true, new Row(), this);
+
+    row.batch_ = [];
+    return row;
+  }
+
+  var self = this;
+  this.batch_ = [];
+
+  fn(function(err) {
+    var requests = self.batch_;
+
+    delete self.batch_;
+
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    self.parent.mutate(requests, callback);
+  });
+};
+
+/**
+ * Mutates a row atomically based on the output of a filter. Depending on
+ * whether or not any results are yielded, either the `onMatch` or `onNoMatch`
+ * callback will be executed.
+ *
+ * @param {object} filter - Filter ot be applied to the contents of the row.
+ * @param {?function} onMatch - Function containing mutations to be ran if a
+ *     match is found.
+ * @param {module:bigtable/row} onMatch.batch - Object used to run mutations in
+ *     the event of a match.
+ * @param {function=} onNoMatch - Function containing mutations to be ran if no
+ *     matches are found.
+ * @param {module:bigtable/row} onNoMatch.batch - Object used to run mutations
+ *     in the event no matches.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this
+ *     request.
+ * @param {boolean} callback.matched - Whether a match was found or not.
+ *
+ * @example
+ * var callback = function(err, matched) {
+ *   if (!err) {
+ *     // `matched` will let us know if a match was found or not.
+ *   }
+ * };
+ *
+ * var filter = {
```

Comment:
On the Java side, we use Chain as filters "ANDed" together and Interleave as "ORed."

If you want a row that meets:

(filter1 AND filter2) OR filter.family('hi') OR (filter3, filter4)

you'd have
interleave(
  chain(filter1, filter2),
  filter.family('hi'),
  chain(filter3, filter4)
)

It's a bit mind bending coming from SQL.  It's almost like a combination of a SELECT clause and a WHERE clause combined in a super strange way.

This isn't a very intuitive model, for things like filter.family('hi') which seem to belong in the "SELECT".  Bigtable has that API because of the sparse nature of Bigtable rows.  families or columns may simply not exist in a row.  

Users usually want
SELECT family(h1)
FROM ..
WHERE (filter1 AND filter2) OR (filter3 AND filter4)

I'm not sure how to cleanly express that in a meaningful API.  The HBase approach adopted by the Java Bigtable implementation leads to all sorts of weird corner cases.

If you'd like to discuss this in more depth, we can discuss this in more depth off line.